vcsclient/fs.go

```diff
@@ -200,18 +200,32 @@ func GetFileWithOptions(fs vfs.FileSystem, path string, opt GetFileOptions) (*Fi
 // readDir uses the passed vfs.FileSystem to read from starting at the base path.
 // If recurseSingleSubfolder is true, it will descend and include sub-folders
 // with a single sub-folder inside. first should always be set to true, other values are used internally.
-func readDir(fs vfs.FileSystem, base string, recurseSingleSubfolder bool, first bool) ([]*TreeEntry, error) {
+func readDir(fs vfs.FileSystem, base string, recurseSingleSubfolder int, first bool) ([]*TreeEntry, error) {
 	entries, err := fs.ReadDir(base)
 	if err != nil {
 		return nil, err
 	}
-	if recurseSingleSubfolder && !first && !singleSubDir(entries) {
+	if recurseSingleSubfolder > 0 && !first && !singleSubDir(entries) {
 		return nil, nil
 	}
 	te := make([]*TreeEntry, len(entries))
 	for i, fi := range entries {
 		te[i] = newTreeEntry(fi)
-		if fi.Mode().IsDir() && recurseSingleSubfolder {
+	}
+
+	if recurseSingleSubfolder > 0 {
+		dirEntries := make(map[int]os.FileInfo)
+		for i, fi := range entries {
+			// expand at most recurseSingleSubfolder subdirectories
+			if len(dirEntries) > recurseSingleSubfolder {
```

Comment:
Nitpicky, but this should be `>=`, else `recurseSingleSubfolder+1` directories will be recursed.